iree/task/task.c

```diff
@@ -0,0 +1,793 @@
+  out_task->header.flags |= IREE_TASK_FLAG_DISPATCH_INDIRECT;
+  out_task->workgroup_count.ptr = workgroup_count_ptr;
+}
+
+void iree_task_dispatch_issue_sliced(iree_task_dispatch_t* dispatch_task,
+                                     iree_task_pool_t* slice_task_pool,
+                                     iree_task_submission_t* pending_submission,
+                                     iree_task_post_batch_t* post_batch) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // Mark the dispatch as having been issued; the next time it retires it'll be
+  // because all work has completed.
+  dispatch_task->header.flags |= IREE_TASK_FLAG_DISPATCH_RETIRE;
+
+  // Fetch the workgroup count (directly or indirectly).
+  // By the task being ready to execute we know any dependencies on the
+  // indirection buffer have been satisfied and its safe to read.
+  uint32_t workgroup_count[3];
+  if (dispatch_task->header.flags & IREE_TASK_FLAG_DISPATCH_INDIRECT) {
+    memcpy(workgroup_count, dispatch_task->workgroup_count.ptr,
+           sizeof(workgroup_count));
+  } else {
+    memcpy(workgroup_count, dispatch_task->workgroup_count.value,
+           sizeof(workgroup_count));
+  }
+
+#if IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_INSTRUMENTATION
+  char xyz_string[32];
+  int xyz_string_length =
+      snprintf(xyz_string, IREE_ARRAYSIZE(xyz_string), "%ux%ux%u",
+               workgroup_count[0], workgroup_count[1], workgroup_count[2]);
+  IREE_TRACE_ZONE_APPEND_TEXT_STRING_VIEW(z0, xyz_string, xyz_string_length);
+#endif  // IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_INSTRUMENTATION
+
+  // Divide up all tiles into slices, our finest-granularity scheduling task.
+  const uint32_t tiles_per_slice_x = IREE_TASK_DISPATCH_TILES_PER_SLICE_X;
+  const uint32_t tiles_per_slice_y = IREE_TASK_DISPATCH_TILES_PER_SLICE_Y;
+  const uint32_t tiles_per_slice_z = IREE_TASK_DISPATCH_TILES_PER_SLICE_Z;
+  uint32_t slice_count_x = iree_max(1, workgroup_count[0] / tiles_per_slice_x);
+  uint32_t slice_count_y = iree_max(1, workgroup_count[1] / tiles_per_slice_y);
+  uint32_t slice_count_z = iree_max(1, workgroup_count[2] / tiles_per_slice_z);
+
+  // Compute how many slices each worker will process.
+  uint32_t slice_count = slice_count_x * slice_count_y * slice_count_z;
+  iree_host_size_t worker_count = iree_task_post_batch_worker_count(post_batch);
+  uint32_t slices_per_worker = iree_max(1, slice_count / worker_count);
+
+  // Randomize starting worker.
+  iree_host_size_t worker_offset = iree_task_post_batch_select_worker(
+      post_batch, dispatch_task->header.affinity_set);
+  iree_host_size_t worker_index = worker_offset;
+
+  // TODO(benvanik): rework this with some science. For now we just iteratively
+  // divide up the space from outer->inner scheduling dimension, but ideally
+  // we'd use some fun cray-style torus scheduling or hilbert curve magic to
+  // try to ensure better locality using worker constructive sharing masks.
```

Comment:
relating to one of my other comments about a max of 64 workers - I'm wondering how much impact this sort of topological scheduling will matter on consumer systems